### ISSUE TITLE
After replacing a disk from the pool, every import starts a resilver?!

### ISSUE THREAD

**Issue by DurvalMenezes** [zfsonlinux/zfs#6813]
### System information

Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       | Devuan GNU/Linux (Debian systemd-less fork)
Distribution Version    | 1.0 (jessie)
Linux Kernel                 | 4.9.0-0.bpo.3-amd64 (from Debian Backports)
Architecture                 | x86_64
ZFS Version                  | 0.6.5.9-2~bpo8+1 (also from  Debian Backports)
SPL Version                  | 0.6.5.9-1~bpo8+1 (ditto)

  
### Describe the problem you're observing
Every time I import this pool (after having exported it previously with no issues), ZoL immediately starts a resilver. Waiting for the resilver to finish and then exporting and reimporting the pool again results in a new resilver, always with the same device being resilvered (ZFS_ARCHIVE_003B2 on the output below), this repeats at every import ad nauseam (I've gone through that 6 times already).

I tried detaching this device from its mirror vdev, then "zpool labelclear"ing it, then attaching it back up and waiting for the resilver to finish. To no avail: when I export and reimport the pool back, again ZoL starts a new resilvering.

FWIW, this disk (ZFS_ARCHIVE_003B2) was previously "zpool replaced" into the pool in order to replace a previous disk which was presenting hard read errors; the replace never finished, even after waiting hours, it still showed "replacing". I was able to export the pool and then detach the previous device and
attach the new one in its place. Could this "never finished replace" be the source of this issue? Exporting and reimporting the pool *without* this device
(ie, after detaching and before reattaching it) does *not* start a resilver in the pool, so definitely the problem is connected to this device.

Also, this disk size is 3TB, while the previous disk it replaced, as well as its pair in the mirror (ZFS_ARCHIVE_003A2 below) are both 2TB disks.
Again, could this be the cause of the problem?

The "3 data errors"shown at the end of the "zpool status" commands below are ancient (more than 1 year ago), from the time the pool was used to copy files off an ECC-less machine which later was diagnosed with RAM issues (the pool has been, since before the current issue, connected to another machine, which has ECC RAM). 

As a final note, all the disk vdevs in this pool are actually LUKS devices running on top of the physical disks -- I've been doing this since forever and I'm pretty sure this has nothing to do with the problem.

### Describe how to reproduce the problem

Just export and reimport the pool, eg:
```
  zpool status pool2015011801
                      pool: pool2015011801
                     state: ONLINE
                    status: One or more devices has experienced an error resulting in data
                            corruption.  Applications may be affected.
                    action: Restore the file in question if possible.  Otherwise restore the
                            entire pool from backup.
                       see: http://zfsonlinux.org/msg/ZFS-8000-8A
                      scan: resilvered 1.21T in 9h4m with 3 errors on Thu Nov  2 14:13:49 2017
                    config:
                            NAME                   STATE     READ WRITE CKSUM
                            pool2015011801         ONLINE       0     0     3
                              mirror-0             ONLINE       0     0     0
                                ZFS_ARCHIVE_003A1  ONLINE       0     0     0
                                ZFS_ARCHIVE_003B1  ONLINE       0     0     0
                              mirror-1             ONLINE       0     0     6
                                ZFS_ARCHIVE_003A2  ONLINE       0     0     6
                                ZFS_ARCHIVE_003B2  ONLINE       0     0     6
                    errors: 3 data errors, use '-v' for a list

                zpool export pool2015011801

                zpool import -d /dev/mapper
                       pool: pool2015011801
                         id: 7358349082112742918
                      state: ONLINE
                     status: The pool is formatted using a legacy on-disk version.
                     action: The pool can be imported using its name or numeric identifier, though
                            some features will not be available without an explicit 'zpool upgrade'.
                     config:
                            pool2015011801         ONLINE
                              mirror-0             ONLINE
                                ZFS_ARCHIVE_003A1  ONLINE
                                ZFS_ARCHIVE_003B1  ONLINE
                              mirror-1             ONLINE
                                ZFS_ARCHIVE_003A2  ONLINE
                                ZFS_ARCHIVE_003B2  ONLINE

                zpool import -d /dev/mapper pool2015011801

                zpool status pool2015011801
                      pool: pool2015011801
                     state: ONLINE
                    status: One or more devices is currently being resilvered.  The pool will
                            continue to function, possibly in a degraded state.
                    action: Wait for the resilver to complete.
                      scan: resilver in progress since Thu Nov  2 14:39:32 2017
                        336M scanned out of 4.31T at 14.6M/s, 85h53m to go
                        185M resilvered, 0.01% done
                    config:
                            NAME                   STATE     READ WRITE CKSUM
                            pool2015011801         ONLINE       0     0     0
                              mirror-0             ONLINE       0     0     0
                                ZFS_ARCHIVE_003A1  ONLINE       0     0     0
                                ZFS_ARCHIVE_003B1  ONLINE       0     0     0
                              mirror-1             ONLINE       0     0     0
                                ZFS_ARCHIVE_003A2  ONLINE       0     0     0
                                ZFS_ARCHIVE_003B2  ONLINE       0     0     0  (resilvering)
                    errors: 3 data errors, use '-v' for a list
```
### Include any warning/errors/backtraces from the system logs

Syslog records nothing relevant, only zed logging the resilver right after every "zpool import":

```
Nov  2 14:39:28 redacted zed[4149]: Finished "all-syslog.sh" eid=63 pid=14120 exit=0
Nov  2 14:39:32 redacted zed[4149]: Invoking "all-syslog.sh" eid=64 pid=14135
Nov  2 14:39:32 redacted zed: eid=64 class=resilver.start pool=pool2015011801
Nov  2 14:39:32 redacted zed[4149]: Finished "all-syslog.sh" eid=64 pid=14135 exit=0
```

**Comment by DurvalMenezes** [comment]
Perhaps this is related to that other issue: #5528.

I will try [the patch suggested there](https://github.com/zfsonlinux/zfs/issues/5528#issuecomment-278014557) and come back to report.

**Comment by behlendorf** [comment]
@DurvalMenezes were you able to try the patch in #5528?  That should resolve your resilver on import issue and leave you with a few damaged files which cannot be repaired.

**Comment by DurvalMenezes** [comment]
@behlendorf: haven't been able to try the patch yet (going through some ZFS-unrelated trouble on another machine). 

Can you please clarify your "and leave you with a few damaged files which cannot be repaired" comment? I thought that importing the pool on a system with #5228 applied would just make ZFS ignore these 3 data errors and finish the resilver,
flagging it as done (and therefore allow me to work with that pool

**Comment by DurvalMenezes** [comment]
Issue closed above by operator/touchscreen error  (finger swiped over the "close issue" button, accidentally activating it) so I immediately reopened. Sorry about the confusion.

**Comment by behlendorf** [comment]
Yes, that's right.  But you'll need to make sure you also enable the module option before import since it's disabled by default.  The reason the resilver keeps restarting is because it's aware there are regions of the disk, which map to a few files in your case, which cannot not be rebuilt.  The only way to clear those errors will be to remove the damaged files.

**Comment by DurvalMenezes** [comment]
Folks, just reporting back (sorry for the delay): I applied the patch from #5228 on top of zfs-0.7.8 (had to apply manually as the surrounding code had changed  enough for `patch` to no longer be able to apply cleanly), recompiled, reinstalled and rebooted to reload the zfs module cleanly; I then changed its controller parameter to activate the code in the patch via  `echo 1 >/sys/module/zfs/parameters/zfs_scan_ignore_errors` and imported the pool; sure enough, the resilver started once more but, once it was over. it was *really* over: it didn't start a new resilver of its own account. 

So, I have my pool again in an usable state. I have further comments on the that patch, which I will post on its issue above, and will close the present issue now.
